contracts/account/utils/draft-ERC7579Utils.sol

```diff
@@ -169,12 +172,34 @@ library ERC7579Utils {
     }
 
     /// @dev Decodes a batch of executions. See {encodeBatch}.
+    ///
+    /// NOTE: This function runs some checks and will throw a {ERC7579DecodingError} if the input is not properly formatted.
     function decodeBatch(bytes calldata executionCalldata) internal pure returns (Execution[] calldata executionBatch) {
-        assembly ("memory-safe") {
-            let ptr := add(executionCalldata.offset, calldataload(executionCalldata.offset))
-            // Extract the ERC7579 Executions
-            executionBatch.offset := add(ptr, 32)
-            executionBatch.length := calldataload(ptr)
+        unchecked {
+            uint256 bufferLength = executionCalldata.length;
+
+            // Check executionCalldata is not empty
+            if (bufferLength < 32) revert ERC7579DecodingError();
+
+            // Get the offset of the array, and check its valid
+            uint256 offset = uint256(bytes32(executionCalldata[0:32]));
+
+            // Array length is between offset and offset + 32. We check that this is in the buffer
+            // Since we know executionCalldata is at least 32, we can subtract with no overflow risk.
+            if (offset > bufferLength - 32) revert ERC7579DecodingError();
+
+            // Get the array length
+            uint256 arrayLength = uint256(bytes32(executionCalldata[offset:offset + 32]));
+            if (arrayLength > type(uint64).max) revert ERC7579DecodingError();
+
+            // Get the array as a bytes slice, and check its is long enough
+            bytes calldata executionArray = executionCalldata[offset + 32:];
+            if (executionArray.length < arrayLength * 96) revert ERC7579DecodingError();
```

Comment:
I think the check for the struct's static-sized elements should be performed by solidity lazily, as you access each `Execution` struct. To make this match the loose ABI-decoder exactly, I think at this stage the only necessary check would be:
```solidity
if (executionArray.length < arrayLength * 32)
```

So, essentially just checking that the struct pointers would still fit in the available space. To illustrate, here is a non-standard, but valid, encoding of 3 `Execution`s that would fail this check, but succeed in `abi.decode(...)` or in the built-in ABI decoder at the top of a function:

```
0x
0000000000000000000000000000000000000000000000000000000000000020    // Offset of `Execution[]`
0000000000000000000000000000000000000000000000000000000000000003    // Length of `Execution[]`
0000000000000000000000000000000000000000000000000000000000000060    // Offset of `executions[0]`
0000000000000000000000000000000000000000000000000000000000000060    // Offset of `executions[1]` (same as `executions[0]`)
0000000000000000000000000000000000000000000000000000000000000060    // Offset of `executions[2]` (same as `executions[0]`)
000000000000000000000000dcaccbf48225b2b2af81351fe52fbd0edf02adff 	// `executions[0].to`
0000000000000000000000000000000000000000000000000000000000000001	// `executions[0].value`
0000000000000000000000000000000000000000000000000000000000000060	// Offset of `executions[0].data`
0000000000000000000000000000000000000000000000000000000000000000	// Length of `executions[0].data`
```

It has all 3 of the offsets point to the same location in calldata, making the total size 7 words, with is less than the expected 9 words.

Either way, it's fine if you keep it as-is, just wanted to flag that this check is ever so slightly too strict.

Reply:
fixed in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5353/commits/6aaa56acdebea0c2b113552b3a287619f0f94406

Reply:
Confirmed with the ABI spec that `arrayLength * 32` is the correct value.